ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java

```diff
@@ -297,7 +299,8 @@ private void takeSnapshot() {
             CommitInfoProto::getCommitIndex);
         purgeIndex = LongStream.concat(LongStream.of(i), commitIndexStream).min().orElse(i);
       }
-      raftLog.purge(purgeIndex);
+      final long purgeIndexWithPreserve = Math.max(0L, purgeIndex - purgePreserveLogNum);
+      raftLog.purge(purgeIndexWithPreserve);
```

Comment:
Let's change it in RaftLogBase.purge(..) so that the new code will be next to the purgeGap code.  We may add a `purgePreservation` field and use it as below
```
+++ b/ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java
@@ -301,6 +301,10 @@ public abstract class RaftLogBase implements RaftLog {
 
   @Override
   public final CompletableFuture<Long> purge(long suggestedIndex) {
+    if (purgePreservation > 0) {
+      final long currentIndex = getNextIndex() - 1;
+      suggestedIndex = Math.min(suggestedIndex, currentIndex - purgePreservation);
+    }
     final long lastPurge = purgeIndex.get();
     if (suggestedIndex - lastPurge < purgeGap) {
       return CompletableFuture.completedFuture(lastPurge);
```

Reply:
Done